modules/utxo-lib/src/bitgo/parseInput.ts

```diff
@@ -0,0 +1,418 @@
+  signatures: [Buffer];
+  publicKeys: [Buffer];
+  pubScript?: Buffer;
+}
+
+export interface ParsedSignatureScript2Of3 extends ParsedSignatureScript {
+  scriptType: 'p2sh' | 'p2shP2wsh' | 'p2wsh';
+  inputClassification: 'scripthash' | 'witnessscripthash';
+  publicKeys: [Buffer, Buffer, Buffer];
+  signatures:
+    | [Buffer, Buffer] // fully-signed transactions with signatures
+    /* Partially signed transactions with placeholder signatures.
+       For p2sh, the placeholder is OP_0 (number 0) */
+    | [Buffer | 0, Buffer | 0, Buffer | 0];
+  pubScript: Buffer;
+}
+
+/**
+ * Keypath spends only have a single pubkey and single signature
+ */
+export interface ParsedSignatureScriptTaprootKeyPath extends ParsedSignatureScript {
+  scriptType: 'p2tr';
+  inputClassification: 'taproot';
+  publicKeys: [Buffer];
+  signatures: [Buffer];
+  pubScript: Buffer;
+}
+
+/**
+ * Taproot Scriptpath spends are more similar to regular p2ms spends and have two public keys and
+ * two signatures
+ */
+export interface ParsedSignatureScriptTaprootScriptPath extends ParsedSignatureScript {
+  scriptType: 'p2tr';
+  inputClassification: 'taproot';
+  publicKeys: [Buffer, Buffer];
+  signatures: [Buffer, Buffer];
+  controlBlock: Buffer;
+  /** Indicates the level inside the taptree. */
+  scriptPathLevel: number;
+  pubScript: Buffer;
+}
+
+export type ParsedSignatureScriptTaproot = ParsedSignatureScriptTaprootKeyPath | ParsedSignatureScriptTaprootScriptPath;
+
+type DecompiledScript = Array<Buffer | number>;
+
+/**
+ * Static script elements
+ */
+type ScriptPatternConstant =
+  | 'OP_0'
+  | 'OP_1'
+  | 'OP_2'
+  | 'OP_3'
+  | 'OP_CHECKMULTISIG'
+  | 'OP_CHECKSIG'
+  | 'OP_CHECKSIGVERIFY';
+
+/**
+ * Script elements that can be captured
+ */
+type ScriptPatternCapture = ':pubkey' | ':signature' | ':buffer' | { ':script': ScriptPatternElement[] };
+
+type ScriptPatternElement = ScriptPatternConstant | ScriptPatternCapture;
+
+/**
+ * Result for a successful script match
+ */
+type MatchResult = {
+  ':pubkey': Buffer[];
+  ':buffer': Buffer[];
+  ':signature': (Buffer | 0)[];
+  ':script': { buffer: Buffer; match: MatchResult }[];
+};
+
+/**
+ * @param script
+ * @param pattern
+ * @return MatchResult if script matches pattern. The result will contain the matched values.
+ */
+function matchScript(script: DecompiledScript, pattern: ScriptPatternElement[]): MatchResult | undefined {
+  /**
+   * Match a single script element with a ScriptPatternElement
+   */
+  function matchElement(e: Buffer | number, p: ScriptPatternElement): boolean | MatchResult {
+    switch (p) {
+      case 'OP_0':
+        return e === opcodes.OP_0 || (Buffer.isBuffer(e) && e.length === 0);
+      case 'OP_1':
+      case 'OP_2':
+      case 'OP_3':
+      case 'OP_CHECKMULTISIG':
+      case 'OP_CHECKSIG':
+      case 'OP_CHECKSIGVERIFY':
+        return e === opcodes[p];
+      case ':pubkey':
+        return Buffer.isBuffer(e) && (e.length === 32 || e.length === 33);
```

Comment:
Wonder if we should split pubkey and xonlypubkey as they do in most of the bips.